feat: enhance download pause/resume with progress persistence

Gauvino opened the pull request:
# 📦 Pull Request

## 🔖 Summary

Enhanced download pause/resume functionality with comprehensive progress persistence and platform-specific handling. Implements reliable pause/resume for Android while temporarily disabling iOS functionality due to dependency limitations.

## 🏷️ Ticket / Issue

- Fixes broken download pause/resume functionality that was continuing downloads in background
- Related to iOS background task zombie process issues identified in dependency research
- https://github.com/kesha-antonov/react-native-background-downloader/issues/26
- https://github.com/kesha-antonov/react-native-background-downloader/issues/24

## 🛠️ What's Changed

- **Type:** feat
- **Scope:** downloads
- **Short summary:** Enhanced download pause/resume with progress persistence and iOS platform limitations handling

## 📋 Details

### Enhanced Download State Management

This PR significantly improves the download pause/resume functionality by:

1. **Extended Progress Tracking**: Added `pausedProgress`, `pausedBytes`, and `pausedAt` fields to download items for comprehensive state persistence
2. **Platform-Specific Implementation**:
   - **Android**: Full pause/resume functionality with reliable progress tracking
   - **iOS**: Temporarily disabled with preserved code structure for future reactivation
3. **Enhanced Error Handling**: Comprehensive logging and graceful error recovery throughout the pause/resume lifecycle

### iOS Dependency Issues Addressed

Research identified critical limitations in the `@kesha-antonov/react-native-background-downloader` dependency:

- **Root Cause**: iOS NSURLSession background tasks become "zombie" processes that continue downloading even after pause commands
- **System-Level Issue**: iOS background downloads are designed to persist at the OS level, making programmatic pause control unreliable
- **Multiple Task Interference**: Zombie tasks cause duplicate downloads and conflicting progress updates

### Technical Implementation

- **Progress Persistence**: Download state properly preserved across pause/resume cycles
- **Platform Detection**: Clean separation using `Platform.OS` checks
- **Future-Ready Code**: All iOS implementation preserved in comments with TODO markers
- **Enhanced Logging**: Comprehensive debug information for troubleshooting

### ⚠️ Breaking Changes

None - existing download functionality remains unchanged, only pause/resume behavior is enhanced.

### 🔐 Security & Privacy Impact

None - no new permissions, data access, or security considerations introduced.

### ⚡ Performance Impact

- **Positive**: Eliminates multiple simultaneous zombie downloads on iOS
- **Neutral**: Android performance unchanged with enhanced state tracking
- **Memory**: Minimal impact from additional progress fields

### 🖼️ Screenshots / GIFs (if UI)

N/A - This is primarily backend functionality enhancement.

## ✅ Checklist

- [x] I've read the [contribution guidelines](CONTRIBUTING.md)
- [x] Code follows project style and passes lint/format (`bun` scripts)
- [x] Type checks pass (tsc/biome/etc.)
- [x] Docs updated (README/ADR/usage/API) - Code comments updated
- [x] No secrets/credentials included; env vars documented
- [x] Release notes/CHANGELOG entry added (if applicable)
- [x] Verified locally that changes behave as expected

## 🔍 Testing Instructions

### Setup

1. `git fetch origin pull/1034/head:fix-download-pause && git checkout fix-download-pause`
2. Install deps: `bun install`
3. Start app: `bun run android` (for Android testing) or `bun run ios` (for iOS testing)

### Android Testing (Full Functionality)

1. Verification steps:
   - [ ] Start a download
   - [ ] Pause the download - should stop immediately with progress preserved
   - [ ] Resume the download - should continue from exact previous progress
   - [ ] Verify no duplicate downloads in background
   - [ ] Check logs show proper pause/resume events

### iOS Testing (Disabled Functionality)

1. Verification steps:
   - [ ] Start a download - should work normally
   - [ ] Attempt to pause - should show error message: "Pause functionality is currently disabled on iOS"
   - [ ] Verify download continues to completion normally
   - [ ] Check logs show warning about disabled iOS functionality

### Edge Cases

- [ ] Multiple downloads pausing/resuming simultaneously
- [ ] App backgrounding during pause state
- [ ] Network connectivity changes during pause/resume
- [ ] Large file downloads with frequent pause/resume cycles

## ⚙️ Deployment Notes

- **Android**: No deployment considerations - functionality works out of the box
- **iOS**: Pause buttons should handle the error gracefully and show appropriate user messaging
- **Dependencies**: No new dependencies added, existing `@kesha-antonov/react-native-background-downloader` version maintained

## 📝 Additional Notes

### Future iOS Reactivation Path

When the underlying dependency limitations are resolved, iOS functionality can be restored by:

1. Removing the `Platform.OS === "ios"` check in `pauseDownload()` and `resumeDownload()`
2. Uncommenting the preserved iOS implementation code (marked with TODO comments)
3. Testing with the enhanced zombie task cleanup sequences

### Dependency Research References

- **GitHub Issue https://github.com/kesha-antonov/react-native-background-downloader/issues/26**: Zombie task problem in `@kesha-antonov/react-native-background-downloader`
- **iOS NSURLSession Documentation**: Background download architecture limitations
- **Platform Differences**: Android DownloadManager vs iOS NSURLSession behavior patterns

### Code Architecture Benefits

- **Maintainability**: Clean platform separation with preserved implementation
- **User Experience**: Prevents broken functionality on iOS while maintaining full Android support
- **Developer Experience**: Clear error messages and comprehensive logging for debugging
- **Future Compatibility**: Ready for reactivation when dependency issues are resolved

This implementation ensures a reliable download experience across platforms while maintaining a clear path for future improvements when the underlying iOS limitations can be addressed.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added download diagnostics export and exposed diagnostics API for troubleshooting.

* **Bug Fixes**
  * Improved download reliability: prevents duplicate tasks, reserves queue slots, preserves progress on resume, caps visible progress at 99%, refines speed estimates, hardens pause/resume flows, and logs previously unhandled promise errors.

* **Style**
  * Moved DownloadCard action buttons to the bottom-right.

* **Chores**
  * Made native build bridging header path dynamic.

* **i18n**
  * Localized Play/Video/Audio/Subtitle, alerts, and download toasts.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Gauvino committed:
feat: enhance download pause/resume with progress persistence

Improves download management by preserving progress state when pausing and resuming downloads.

Adds pausedAt, pausedProgress, and pausedBytes fields to track download state during interruptions. Implements robust pause handling with fallback to stop operation if pause fails. Enhances resume logic to restore previous progress for interrupted downloads and handle failed tasks gracefully.

Prevents duplicate download processes for the same item and adds proper error handling for background download operations.

Gauvino commented:
@coderabbitai full review

fredrikburmester commented:
It does not seem like this worked on iOS at all, sadly!

Simon-Eklundh commented on app/_layout.tsx:
```diff
@@ -398,11 +398,15 @@ function Layout() {
-    BackGroundDownloader.checkForExistingDownloads();
+    BackGroundDownloader.checkForExistingDownloads().then(
+      (_tasks: any[]) => {},
+    );
```
check for existing downloads and then do nothing with them?

Simon-Eklundh commented on providers/DownloadProvider.tsx:
```diff
@@ -176,21 +176,46 @@ function useDownloadProvider() {
-          if (progress >= 100) {
-            progress = 99;
+
+          if (p.pausedProgress !== undefined && p.pausedProgress > 0) {
```
this should be worth taking into account considering the possibility to queue downloads

Simon-Eklundh commented on providers/DownloadProvider.tsx:
```diff
@@ -176,21 +176,46 @@ function useDownloadProvider() {
+            const totalBytesDownloaded = p.pausedBytes
+              ? p.pausedBytes + task.bytesDownloaded
+              : task.bytesDownloaded;
+            const speed = calculateSpeed(p, task.bytesDownloaded);
```
This might make sense, don't know for right now as I seem to see pausedbytes being the same as bytesDownloaded....

probably braindamage but...

Simon-Eklundh commented on providers/DownloadProvider.tsx:
```diff
@@ -176,21 +176,46 @@ function useDownloadProvider() {
+            };
+          } else {
+            if (estimatedSize > 0) {
+              progress = (100 / estimatedSize) * task.bytesDownloaded;
```
these brackets are unnecessary

and the math breaks my brain :D

Simon-Eklundh commented on providers/DownloadProvider.tsx:
```diff
@@ -795,8 +826,51 @@ function useDownloadProvider() {
+        if (
+          verifyTask &&
+          verifyTask.state !== "PAUSED" &&
+          verifyTask.state !== "paused"
```
```suggestion
          verifyTask &&
          verifyTask.state.toLowerCase() !== "paused"
```

Simon-Eklundh commented on providers/DownloadProvider.tsx:
```diff
@@ -795,8 +826,51 @@ function useDownloadProvider() {
+        updateProcess(id, {
+          status: "paused",
+          progress: currentProgress,
+          bytesDownloaded: currentBytes,
+          pausedAt: new Date(),
+          pausedProgress: currentProgress,
+          pausedBytes: currentBytes,
```
what's the difference between paused bytes and downloaded bytes? XD

Gauvino replied:
pausedBytes is the persisted snapshot saved when you pause (durable). bytesDownloaded is the volatile counter from the current native/background task session. On resume we use pausedBytes + task.bytesDownloaded (or the partial file size) so the UI doesn’t jump back to 0%.

Simon-Eklundh replied:
Ah, I see!

Simon-Eklundh reviewed (commented):
some minor nits and similar, and I haven't tested it yet

Simon-Eklundh commented:
tested now; you need to merge with develop, and starting downloads breaks the top bar again (like the "downloads" button on the home screen

and it didn't seem like the remux amount of downloads setting was followed

Gauvino committed:
Merge branch 'develop' into fix-download-pause

Gauvino committed:
fix(download): improve error handling and pause/resume reliability

Improves background download robustness by adding proper error handling and fallback mechanisms for pause/resume operations.

Enhances pause functionality with graceful fallbacks when native pause fails, ensuring downloads can be properly stopped and resumed later.

Adds better size estimation using bitrate and runtime when media size is unavailable, preventing download progress calculation errors.

Fixes race conditions in concurrent download management by immediately reserving slots when starting queued downloads.

Includes user notifications for duplicate download attempts and improved partial file detection for more accurate resume progress.

Gauvino committed:
refactor: replace inline typecheck with dedicated script

Replaces the inline TypeScript checking command with a dedicated Node.js script to improve maintainability and handle complex filtering logic more effectively.

Updates jellyseerr submodule to latest commit.

Gauvino committed:
feat: add session-based speed calculation for downloads

Implements session-only speed tracking to provide more accurate download speed measurements when resuming paused downloads.

Adds lastSessionBytes and lastSessionUpdateTime fields to track bytes downloaded since the last resume operation. Updates speed calculation logic to prefer session-based deltas over total download deltas, falling back to the original method for compatibility.

Preserves session tracking data during pause and resume operations to ensure speed calculations reflect current session performance rather than overall download history.

Gauvino committed:
feat: add download diagnostics functionality and fix hardcoded bridging header path

Introduces diagnostic capabilities for troubleshooting download issues by collecting JavaScript processes and native task information into exportable JSON files.

Resolves hardcoded project name reference in Xcode configuration that prevented proper builds for projects with different names by dynamically using the detected project name for bridging header paths.

Gauvino committed:
Merge branch 'develop' into fix-download-pause

Gauvino committed:
fix: move action buttons to bottom right corner

Repositions download card action buttons from top-right to bottom-right corner to improve UI layout and user accessibility.

Gauvino committed:
feat: add RN background downloader plugin

Enables background downloading functionality by integrating the React Native background downloader plugin into the Expo configuration.

Gauvino committed:
fix: remove unecessary change

Gauvino committed:
fix: handle download start failures with error handling

Adds proper error handling to catch failures when starting downloads, preventing unhandled promise rejections.

Updates download status to error state and displays user-friendly toast notification when download initialization fails.

Gauvino committed:
Merge branch 'develop' into fix-download-pause

Gauvino committed:
fix: improve download management and iOS pause handling

Replaces duplicate download prevention with automatic replacement to avoid user confusion and improve experience.

Adds platform-specific pause behavior for iOS where pause() is unreliable, forcing tasks to stop instead of continuing in background.

Maintains existing pause logic for Android and other platforms while ensuring consistent download state management across all platforms.

Gauvino committed:
Merge branch 'develop' into fix-download-pause

Gauvino committed:
refactor: improve download progress tracking accuracy

Renames progress constant for better clarity and preserves download state when resuming downloads.

Updates session byte tracking to use actual downloaded bytes instead of conditional logic that could cause inconsistencies.

Ensures downloads can resume from their previous state rather than resetting to zero when restarted.

Gauvino committed:
Merge branch 'develop' into fix-download-pause

Gauvino committed:
fix: improve download resume functionality for iOS

Preserves existing progress when resuming downloads and properly handles byte counting for iOS platform.

Addresses iOS-specific behavior where the downloader restarts from zero by adding previously downloaded bytes to the current progress calculation.

Clears paused state fields when resuming to prevent stale data from affecting subsequent operations.

Simplifies diagnostics logging by removing unused dump function and using standard log utility.

Gauvino committed:
fix: resolve iOS download resume issue with proper progress tracking

- Add resumeOffset field to JobStatus type for iOS resume calculations
- Update resumeDownload to set resumeOffset when resuming paused downloads
- Modify progress callback to use resumeOffset for accurate progress on iOS
- Clear paused state properly while preserving resume progress information

This fixes the 'two downloads' visual bug where resuming on iOS would start from 0% instead of continuing from the paused position.

Gauvino committed:
Revert "fix: resolve iOS download resume issue with proper progress tracking"

This reverts commit 177c4ffe6649e834f5c94b3eb757fe257b158239.

Gauvino committed:
Revert "fix: improve download resume functionality for iOS"

This reverts commit 478935703e83fef18d57e77c6eaa82344b959ab7.

Gauvino committed:
fix: properly pause downloads on iOS to allow resuming

- Change iOS pause logic to use task.pause() instead of task.stop()
- task.stop() cancels the background task preventing resume
- task.pause() suspends the task allowing proper resume functionality
- Add fallback to stop if pause fails
- This should fix the iOS resume issue where downloads restart from beginning

The issue was that on iOS, pausing called task.stop() which cancels the background download task. Once cancelled, the task cannot be resumed. The fix is to use task.pause() which suspends the task, allowing it to be resumed later with the correct progress.

Gauvino committed:
fix(download): improve iOS pause handling and add diagnostics

Simplifies iOS download pause logic by always stopping tasks instead of attempting pause operations, which are unreliable on iOS platform.

Preserves existing download progress when resuming downloads to prevent progress loss during resume operations.

Adds error handling around task stop and completion operations to prevent crashes from native module failures.

Introduces download diagnostics utility for debugging download state issues with processes, native tasks, and focused process information.

Gauvino committed:
fix(download): improve iOS download pause/resume behavior

Enhances download task management on iOS by properly handling suspended tasks and implementing more reliable pause functionality.

For cancellation operations, adds logic to resume suspended tasks before stopping them to ensure proper cleanup.

For pause operations, replaces the previous stop-based approach with proper task suspension, including verification of pause state and fallback mechanisms when pause fails.

Maintains session bookkeeping data to preserve download progress and speed calculations across pause/resume cycles.

Gauvino committed:
Merge branch 'develop' into fix-download-pause

Gauvino committed:
refactor: simplify download pause/resume logic

Removes complex platform-specific pause/resume handling in favor of a unified approach that stops and restarts downloads consistently across iOS and Android.

The previous implementation had extensive error handling and fallback mechanisms for different pause states, which caused unpredictable behavior. The new approach always stops the task cleanly and restarts from the saved progress point, ensuring reliable pause/resume functionality regardless of platform.

Gauvino committed:
Merge branch 'develop' into fix-download-pause

Gauvino committed:
fix(download): improve resume reliability and progress tracking

Prevents duplicate download tasks by cleaning up existing downloads before starting new ones

Enhances progress calculation for resumed downloads by tracking paused bytes and estimating total file size

Adds logging for better debugging of resume operations and includes timing delays to ensure proper cleanup sequencing

Gauvino committed:
fix: Enhanced iOS pause/resume with zombie task cleanup - Based on GitHub issue research, implemented comprehensive iOS fixes

Gauvino commented:
Waiting for #1016 to be merged

Gauvino committed:
feat: disable iOS pause/resume functionality while preserving code

- Temporarily disabled pause/resume functionality on iOS due to background task issues
- All iOS-specific code is commented out with TODO markers for easy reactivation
- Android pause/resume functionality remains fully functional
- Enhanced error messages inform users about iOS limitation
- Code structure preserved for future iOS implementation

To re-enable iOS functionality in the future:
1. Remove platform checks in pauseDownload and resumeDownload functions
2. Uncomment iOS-specific cleanup code blocks
3. Uncomment zombie task detection in updateProgress function

Gauvino committed:
feat: internationalize hardcoded strings across UI components

Replaces hardcoded English strings with translation keys to support localization.

Adds translation keys for common UI elements like "Video", "Audio", "Subtitle", "Play", and error messages. Updates toast notifications and alerts to use proper translation keys with parameters for dynamic content.

Enhances user experience by enabling proper multilingual support throughout the application.

Gauvino committed:
Merge branch 'develop' into fix-download-pause

Gauvino commented:
# Additional Fix: Resolved Missing Translation Keys

While working on this PR, I also identified and fixed several missing translation keys that were causing raw key names to display instead of proper translated text:

## Translation Issues Fixed:
- ✅ Fixed hardcoded "No trailer available" alert in SeriesActions component
- ✅ Replaced hardcoded "Video", "Audio", "Subtitle" titles with translation keys in sessions page
- ✅ Fixed "Failed to start download" toast message in DownloadProvider
- ✅ Added translation support for background download enable/disable toasts
- ✅ Fixed hardcoded "Play" button text in Jellyseerr page
- ✅ Added missing translation keys to `translations/en.json`

## Keys Added:
- `common.no_trailer_available`
- `common.video`, `common.audio`, `common.subtitle`
- `common.play`
- `home.downloads.toasts.failed_to_start_download`
- `home.settings.toasts.background_downloads_enabled/disabled`

This ensures a consistent internationalization experience and prevents raw translation keys from appearing in the UI.

Gauvino committed:
fix: improve error handling and download progress calculation

Replaces empty catch blocks with proper error logging for background download operations to aid in debugging.

Simplifies download progress calculation by using total bytes downloaded directly instead of complex session-based percentage calculations, improving accuracy and maintainability.

Adds missing dependency to React hook dependency array and improves diagnostic logging utility with better default parameter handling.

Gauvino committed:
fix: disable pause/resume buttons on iOS downloads

Prevents pause and resume functionality from being available on iOS platform where these operations are not supported or may cause issues with the download process.

Gauvino committed:
Merge branch 'develop' into fix-download-pause
